src/YHttpWidgetsActionHandler.h

```diff
@@ -46,45 +51,108 @@ class YHttpWidgetsActionHandler : public YHttpHandler
 
     virtual int errorCode() {return _error_code;}
 
-    virtual std::string contentEncoding();
-
-private:
-
-    int _error_code;
+    virtual std::string contentType();
 
     // TODO: move this somewhere else...
-
-    int do_action(YWidget *widget, const std::string &action, struct MHD_Connection *connection, std::ostream& body);
-
     template<typename T>
-    int action_handler(YWidget *widget, std::function<void (T*)> handler_func) {
+    int action_handler( YWidget *widget, std::ostream& body, std::function<void (T*)> handler_func ) {
         if (auto w = dynamic_cast<T*>(widget)) {
             try
             {
                 // allow changing only the enabled widgets, disabled ones
                 // cannot be changed by user from the UI, do not be more powerfull
-                if (handler_func && widget->isEnabled()) handler_func(w);
+                if( !widget->isEnabled() )
+                {
+                    body << "{ error: \"Cannot operate on disabled widget: '" << typeid(*widget).name() << "'\" }" << std::endl;
+                    return MHD_HTTP_UNPROCESSABLE_ENTITY;
+                }
+                if ( handler_func )
+                    handler_func(w);
             }
             // some widgets may throw an exception when setting invalid values
             catch (const YUIException &e)
             {
-                yuiError() << "Error while processing action on widget "
-                    << typeid(*widget).name() << " " << e.what() << std::endl;
+                body << "{ error: \"" << typeid(*widget).name() << " " << e.what() << "\" }" << std::endl;
```

Comment:
Using a literal JSON is fine for static error messages:
```cpp
body << "{ \"error\" : \"Widget not found\" }" << std::endl;
```

But for variable massages which can potentially contain some user input it is not safe. E.g. if the exception text contains a double quote, new line or other special characters then it can break the JSON format. In these cases we would need to escape the message text. But it's easier to simply use the JSON library for that:

```cpp
Json::Value response;
response["error"] = typeid(*widget).name() + " " + e.what();
YJsonSerializer::save(response, body);
```

Reply:
Makes sense, I will fix that while resolving conflicts. Thanks!